urllib3/filepost.py

```diff
@@ -56,39 +53,165 @@ def iter_fields(fields):
+    is either a (filename, data) tuple or just data.
+
+    The data can be a unicode string, an iterator producing strings, or a file-like
+    object. File-like objects are read ``chunk_size`` bytes at a time.
+
+    If no ``boundary`` is specified then a random one is used.
+    """
+    def __init__(self, fields, boundary=None, chunk_size=8192):
+        self.fields = [field for field in iter_field_objects(fields)]
+        self.chunk_size = chunk_size
+        self.boundary = boundary or choose_boundary()
+
+    def get_content_type(self):
+        return 'multipart/form-data; boundary=%s' % self.boundary
+
+    def __len__(self):
+        """
+        Figure out the expected body size by iterating over the fields as if they
+        contained empty files, while accumulating the value file sizes as
+        efficiently as we can.
+        """
+        empty_fields = []
+        size = 0
+        for field in self.fields:
+            name = field._name
+            data = field.data
+            filename = field._filename
+
+            if filename is not None:
+                empty_fields.append((name, (filename, '')))
+            else:
+                empty_fields.append((name, ''))
+
+            if hasattr(data, '__len__'):
+                size += len(encode(data))
+            elif isinstance(data, six.integer_types):
+                size += len(str(data))
+            elif hasattr(data, 'seek'):
+                size += file_size(data)
+            elif hasattr(data, 'read'):
+                size += len(encode(data.read()))  # This is undesired
+            elif hasattr(data, '__iter__'):
+                size += sum(len(encode(chunk)) for chunk in data)  # This is also undesired
+            else:
+                size += len(encode(data))  # Hope for the best
+
+        empty_multipart = MultipartEncoderGenerator(empty_fields, boundary=self.boundary)
+        empty_size = sum(encode(len(chunk)) for chunk in empty_multipart)
+
+        return size + empty_size
+
+    def __iter__(self):
+        for field in self.fields:
+            data = field.data
+            yield encode('--%s\r\n' % (self.boundary))
+            yield encode(field.render_headers())
+
+            if isinstance(data, bytes):
+                yield data
+
+            elif isinstance(data, (str, six.text_type)):
+                yield encode(data)
+
+            elif isinstance(data, six.integer_types):
+                # Handle integers for backwards compatibility
+                yield encode(str(data))
+
+            elif hasattr(data, 'read'):
+                # Stream from a file-like object
+                while True:
+                    chunk = data.read(self.chunk_size)
+                    if not chunk:
+                        break
+                    yield encode(chunk)
+
+            elif hasattr(data, '__iter__'):
+                # Stream from an iterator
+                for chunk in data:
+                    yield encode(chunk)
```

Comment:
I note that there is a lot of assumption that iterators will return text. That's probably not right. The text/bytes handling stuff should probably be factored out. It may be sensible to try to nest iterators here so that we can factor out the encode/decode logic.